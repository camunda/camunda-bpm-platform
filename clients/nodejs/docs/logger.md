# logger

A logger is a simple middleware that logs various events in the client lifecycle. It can be configured using different log-levels.

```js
const { Client, logger } = require("camunda-external-task-client-js");

const client = new Client({
  use: logger,
  baseUrl: "http://localhost:8080/engine-rest"
});
```

## `logger.level(logLevel)`
Returns a logger instance with the configured log level.
```js
const { Client, logger } = require("camunda-external-task-client-js");

const client = new Client({
  use: logger.level('debug'),
  baseUrl: "http://localhost:8080/engine-rest"
});
```

The levels correspond to the npm logging levels:

```JSON
{
  error: 0,
  warn: 1,
  info: 2,
  verbose: 3,
  debug: 4,
  silly: 5
}
```

If you do not specify a log level, `info` will be used.


## `logger.success(text)`

Receives a text and produces a success message out of it.

```js
console.log(logger.success("This is a success message!"));
```

![logger.success](./logger-success.png)



## `logger.error(text)`
Receives a text and produces an error message out of it.

```js
console.log(logger.error("This is an error message!"));
```

![logger.error](./logger-error.png)
