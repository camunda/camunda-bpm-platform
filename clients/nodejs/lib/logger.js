/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import chalk from "chalk";

/**
 * @returns a formatted success message
 */
const success = (message) => `${chalk.green("✓")} ${chalk.green(message)}`;

/**
 * @returns a formatted error message
 */
const error = (message) => `${chalk.red("✖")} ${chalk.red(message)}`;

const levels = {
  error: 0,
  warn: 1,
  info: 2,
  verbose: 3,
  debug: 4,
  silly: 5,
};

/**
 * logs various events from client
 * @param client
 */
const logger = (client, clientLogLevel) => {
  const log = (messageLogLevel, message) => {
    if (!message) {
      console.log(messageLogLevel);
      return;
    }

    if (levels[messageLogLevel] <= clientLogLevel) {
      console.log(message);
    }
  };

  switch (typeof clientLogLevel) {
    case "string":
      clientLogLevel = levels[clientLogLevel];
      break;
    case "number":
      break;
    default:
      clientLogLevel = levels["info"];
      break;
  }

  client.on("subscribe", (topic) => {
    log("info", success(`subscribed to topic ${topic}`));
  });

  client.on("unsubscribe", (topic) => {
    log("info", success(`unsubscribed from topic ${topic}`));
  });

  client.on("poll:start", () => {
    log("debug", "polling");
  });

  client.on("poll:stop", () => {
    log("debug", error("polling stopped"));
  });

  client.on("poll:success", (tasks) => {
    const output = success(`polled ${tasks.length} tasks`);
    log("debug", output);
  });

  client.on("poll:error", (e) => {
    const output = error(`polling failed with ${e}`);
    log("error", output);
  });

  client.on("complete:success", ({ id }) => {
    log("info", success(`completed task ${id}`));
  });

  client.on("complete:error", ({ id }, e) => {
    log("error", error(`couldn't complete task ${id}, ${e}`));
  });

  client.on("handleFailure:success", ({ id }) => {
    log("info", success(`handled failure of task ${id}`));
  });

  client.on("handleFailure:error", ({ id }, e) => {
    log("error", error(`couldn't handle failure of task ${id}, ${e}`));
  });

  client.on("handleBpmnError:success", ({ id }) => {
    log("info", success(`handled BPMN error of task ${id}`));
  });

  client.on("handleBpmnError:error", ({ id }, e) => {
    log("error", error(`couldn't handle BPMN error of task ${id}, ${e}`));
  });

  client.on("extendLock:success", ({ id }) => {
    log("info", success(`handled extend lock of task ${id}`));
  });

  client.on("extendLock:error", ({ id }, e) => {
    log("error", error(`couldn't handle extend lock of task ${id}, ${e}`));
  });

  client.on("unlock:success", ({ id }) => {
    log("info", success(`unlocked task ${id}`));
  });

  client.on("unlock:error", ({ id }, e) => {
    log("error", error(`couldn't unlock task ${id}, ${e}`));
  });

  client.on("lock:success", ({ id }) => {
    log("info", success(`locked task ${id}`));
  });

  client.on("lock:error", ({ id }, e) => {
    log("error", error(`couldn't lock task ${id}, ${e}`));
  });
};

/**
 * Returns Logger with configured log-level
 * @param level
 */
const level = (level) => {
  return function (client) {
    logger(client, level);
  };
};

// export logger & attach to it success and error methods
export default Object.assign(logger, { success, error, level });
