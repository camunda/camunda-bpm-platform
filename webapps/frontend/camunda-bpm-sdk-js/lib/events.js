/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

/**
 * Events handling utility which can be used on
 * any kind of object to provide `on`, `once`, `off`
 * and `trigger` functions.
 *
 * @exports CamSDK.Events
 * @mixin
 *
 * @example
 * var obj = {};
 * Events.attach(obj);
 *
 * obj.on('event:name', function() {});
 * obj.once('event:name', function() {});
 * obj.trigger('event:name', data, moreData, evenMoreData);
 */

var Events = {};

/**
 * Converts an object into array
 * @param  {*} obj
 * @return {Array}
 */
function toArray(obj) {
  var a,
    arr = [];
  for (a in obj) {
    arr.push(obj[a]);
  }
  return arr;
}

/**
 * Returns a function that will be executed
 * at most one time, no matter how often you call it.
 * @param  {Function} func
 * @return {Function}
 */
function once(func) {
  var ran = false,
    memo;
  return function() {
    if (ran) return memo;
    ran = true;
    memo = func.apply(this, arguments);
    func = null;
    return memo;
  };
}

/**
 * Ensure an object to have the needed _events property
 * @param  {*} obj
 * @param  {String} name
 */
function ensureEvents(obj, name) {
  obj._events = obj._events || {};
  obj._events[name] = obj._events[name] || [];
}

/**
 * Add the relevant Events methods to an object
 * @param  {*} obj
 */
Events.attach = function(obj) {
  obj.on = this.on;
  obj.once = this.once;
  obj.off = this.off;
  obj.trigger = this.trigger;
  obj._events = {};
};

/**
 * Bind a callback to `eventName`
 * @param  {String}   eventName
 * @param  {Function} callback
 */
Events.on = function(eventName, callback) {
  ensureEvents(this, eventName);

  this._events[eventName].push(callback);

  return this;
};

/**
 * Bind a callback who will only be called once to `eventName`
 * @param  {String}   eventName
 * @param  {Function} callback
 */
Events.once = function(eventName, callback) {
  var self = this;
  var cb = once(function() {
    self.off(eventName, once);
    callback.apply(this, arguments);
  });
  cb._callback = callback;
  return this.on(eventName, cb);
};

/**
 * Unbind one or all callbacks originally bound to `eventName`
 * @param  {String}   eventName
 * @param  {Function} [callback]
 */
Events.off = function(eventName, callback) {
  ensureEvents(this, eventName);

  if (!callback) {
    delete this._events[eventName];
    return this;
  }

  var e,
    arr = [];
  for (e in this._events[eventName]) {
    if (this._events[eventName][e] !== callback) {
      arr.push(this._events[eventName][e]);
    }
  }
  this._events[eventName] = arr;

  return this;
};

/**
 * Call the functions bound to `eventName`
 * @param  {String} eventName
 * @param {...*} [params]
 */
Events.trigger = function() {
  var args = toArray(arguments);
  var eventName = args.shift();
  ensureEvents(this, eventName);

  var e;
  for (e in this._events[eventName]) {
    this._events[eventName][e](this, args);
  }

  return this;
};

module.exports = Events;
