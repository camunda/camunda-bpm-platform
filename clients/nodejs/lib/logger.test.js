/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { jest } from "@jest/globals";
import logger from "./logger.js";
import events from "events";

describe("logger", () => {
  let client;

  describe("events", () => {
    beforeAll(() => {
      client = new events();
      logger(client, "silly");
    });

    beforeEach(() => {
      jest.spyOn(console, "log").mockImplementation(() => jest.fn());
    });

    afterEach(() => {
      console.log.mockClear();
    });

    it("should log subscribe event", () => {
      // when
      client.emit("subscribe", "some topic");

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log unsubscribe event", () => {
      // when
      client.emit("unsubscribe", "some topic");

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log poll:start event", () => {
      // when
      client.emit("poll:start");

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log poll:stop event", () => {
      // when
      client.emit("poll:stop");

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log poll:success event", () => {
      // when
      client.emit("poll:success", ["task1", "task2"]);

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log poll:error event", () => {
      // when
      client.emit("poll:error");

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log complete:success", () => {
      // when
      client.emit("complete:success", { id: "some task id" });

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log complete:error", () => {
      // when
      client.emit("complete:error", { id: "some task id" }, "some error");

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log handleFailure:success", () => {
      // when
      client.emit("handleFailure:success", { id: "some task id" });

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log handleFailure:error", () => {
      // when
      client.emit("handleFailure:error", { id: "some task id" }, "some error");

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log handleBpmnError:success", () => {
      // when
      client.emit("handleBpmnError:success", { id: "some task id" });

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log handleBpmnError:error", () => {
      // when
      client.emit(
        "handleBpmnError:error",
        { id: "some task id" },
        "some error"
      );

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log extendLock:success", () => {
      // when
      client.emit("extendLock:success", { id: "some task id" });

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log extendLock:error", () => {
      // when
      client.emit("extendLock:error", { id: "some task id" }, "some error");

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log unlock:success", () => {
      // when
      client.emit("unlock:success", { id: "some task id" });

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should log unlock:error", () => {
      // when
      client.emit("unlock:error", { id: "some task id" }, "some error");

      // then
      expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe("config", () => {
    beforeEach(() => {
      client = new events();
      jest.spyOn(console, "log").mockImplementation(() => jest.fn());
    });

    afterEach(() => {
      console.log.mockClear();
    });

    it("should have debug level", () => {
      // when
      logger(client, "debug");

      client.emit("poll:start"); // debug
      client.emit("subscribe", "some topic"); // info
      client.emit("poll:error"); // error

      // then
      expect(console.log).toHaveBeenCalledTimes(3);
    });

    it("should have info level", () => {
      // when
      logger(client);

      client.emit("poll:start"); // debug
      client.emit("subscribe", "some topic"); // info
      client.emit("poll:error"); // error

      // then
      expect(console.log).toHaveBeenCalledTimes(2);
    });

    it("should have error level", () => {
      // when
      logger(client, "error");

      client.emit("poll:start"); // debug
      client.emit("subscribe", "some topic"); // info
      client.emit("poll:error"); // error

      // then
      expect(console.log).toHaveBeenCalledTimes(1);
    });

    it("should use info level on default", () => {
      // when
      logger(client);

      client.emit("poll:start"); // debug
      client.emit("subscribe", "some topic"); // info
      client.emit("poll:error"); // error

      // then
      expect(console.log).toHaveBeenCalledTimes(2);
    });

    it("should allow NPM log level numbers", () => {
      // when
      logger(client, 2); // info

      client.emit("poll:start"); // debug
      client.emit("subscribe", "some topic"); // info
      client.emit("poll:error"); // error

      // then
      expect(console.log).toHaveBeenCalledTimes(2);
    });
  });
});
