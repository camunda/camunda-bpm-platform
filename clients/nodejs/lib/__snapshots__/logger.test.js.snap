// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger events should log complete:error 1`] = `"✖ couldn't complete task some task id, some error"`;

exports[`logger events should log complete:success 1`] = `"✓ completed task some task id"`;

exports[`logger events should log extendLock:error 1`] = `"✖ couldn't handle extend lock of task some task id, some error"`;

exports[`logger events should log extendLock:success 1`] = `"✓ handled extend lock of task some task id"`;

exports[`logger events should log handleBpmnError:error 1`] = `"✖ couldn't handle BPMN error of task some task id, some error"`;

exports[`logger events should log handleBpmnError:success 1`] = `"✓ handled BPMN error of task some task id"`;

exports[`logger events should log handleFailure:error 1`] = `"✖ couldn't handle failure of task some task id, some error"`;

exports[`logger events should log handleFailure:success 1`] = `"✓ handled failure of task some task id"`;

exports[`logger events should log poll:error event 1`] = `"✖ polling failed with undefined"`;

exports[`logger events should log poll:start event 1`] = `"polling"`;

exports[`logger events should log poll:stop event 1`] = `"✖ polling stopped"`;

exports[`logger events should log poll:success event 1`] = `"✓ polled 2 tasks"`;

exports[`logger events should log subscribe event 1`] = `"✓ subscribed to topic some topic"`;

exports[`logger events should log unlock:error 1`] = `"✖ couldn't unlock task some task id, some error"`;

exports[`logger events should log unlock:success 1`] = `"✓ unlocked task some task id"`;

exports[`logger events should log unsubscribe event 1`] = `"✓ unsubscribed from topic some topic"`;
